File: backend/models/Book.js
const mongoose = require('mongoose');

const bookSchema = new mongoose.Schema({
  title: { type: String, required: true, trim: true },
  author: { type: String, required: true, trim: true },
  description: { type: String, required: true },
  imageUrl: { type: String, required: true },
  genre: { type: String, required: true, trim: true },
  summary: { type: String, required: true },
  content: { type: String, default: '' },
  price: { type: Number, required: true, min: 0 },
  rating: { type: Number, default: 0, min: 0, max: 5 },
  totalRatings: { type: Number, default: 0, min: 0 },
  purchaseUrl: { type: String, default: '' },
  pdfUrl: { type: String, default: '' },
  ratings: [
    {
      value: { type: Number, required: true, min: 1, max: 5 },
      date: { type: Date, default: Date.now }
    }
  ]
}, { timestamps: true });

// Get all books with optional genre and sort
bookSchema.statics.getAll = async function (genre = null, sort = 'latest') {
  const query = genre ? { genre } : {};
  let sortObj = { createdAt: -1 };
  switch ((sort || 'latest').toLowerCase()) {
    case 'popular':
      sortObj = { totalRatings: -1, rating: -1, createdAt: -1 };
      break;
    case 'rated':
    case 'highly-rated':
      sortObj = { rating: -1, totalRatings: -1, createdAt: -1 };
      break;
    case 'latest':
    default:
      sortObj = { createdAt: -1 };
  }
  return await this.find(query).sort(sortObj);
};

// Get book by ID (returns Mongoose document)
bookSchema.statics.getById = async function (id) {
  if (!mongoose.Types.ObjectId.isValid(id)) return null;
  return await this.findById(id);
};

// Add rating
bookSchema.methods.addRating = async function (ratingValue) {
  this.ratings.push({ value: ratingValue });
  const totalValue = this.ratings.reduce((sum, r) => sum + r.value, 0);
  this.rating = Math.round((totalValue / this.ratings.length) * 10) / 10; // one decimal
  this.totalRatings = this.ratings.length;
  return await this.save();
};

const Book = mongoose.model('Book', bookSchema);
module.exports = Book;

File: echoreads_mobile/lib/models/book.dart
class Book {
  final String id;
  final String title;
  final String author;
  final String? imageUrl;
  final String? description;
  final String? summary;
  final String? content;
  final double rating;
  final int totalRatings;
  final String? genre;
  final double price;

  Book({
    required this.id,
    required this.title,
    required this.author,
    this.imageUrl,
    this.description,
    this.summary,
    this.content,
    required this.rating,
    required this.totalRatings,
    this.genre,
    required this.price,
  });

  factory Book.fromJson(Map<String, dynamic> json) {
    final dynamic idRaw = json['_id'] ?? json['id'];
    return Book(
      id: idRaw?.toString() ?? '',
      title: json['title'] ?? 'Untitled',
      author: json['author'] ?? 'Unknown',
      imageUrl: json['imageUrl'] as String?,
      description: json['description'] as String?,
      summary: json['summary'] as String?,
      content: json['content'] as String?,
      rating: (json['rating'] is int)
          ? (json['rating'] as int).toDouble()
          : (json['rating'] ?? 0.0).toDouble(),
      totalRatings: (json['totalRatings'] ?? 0) as int,
      genre: json['genre'] as String?,
      price: (json['price'] is int)
          ? (json['price'] as int).toDouble()
          : (json['price'] ?? 0.0).toDouble(),
    );
  }
}
